core/src/main/scala/org/http4s/HttpToken.scala

```diff
@@ -0,0 +1,21 @@
+package org.http4s
+
+import org.http4s.util.{Renderable, Writer}
+
+abstract class HttpToken(val value: String) extends Renderable
+
+final case class HttpProduct(override val value: String, version: Option[String] = None)
+    extends HttpToken(value) {
+  override def render(writer: Writer): writer.type = {
+    writer << value
+    version.foreach { v =>
+      writer << '/' << v
+    }
+    writer
+  }
+}
+
+final case class HttpComment(override val value: String) extends HttpToken(value) {
```

Comment:
Why is a comment a token?  It's a different thing.

```
     token          = 1*tchar
     comment        = "(" *( ctext / quoted-pair / comment ) ")"
```

I guess a sum type is needed for the `product / comment` rule.  I think `HttpToken` is maybe misleading.  `ProductOrComment` comes to mind, but is verbose.  Union types come to mind, but we're not on dotty yet!

> The User-Agent field-value consists of one or more product identifiers, each followed by zero or more comment

When thought of that way, maybe this model becomes useful:

```scala
case class Product(value: String, version: Option[String], comments: List[String])
```

Reply:
Trying to rename it to `Product` leads to quite many conflicts on the code base with `scala.Product`. Maybe we should go back to `ProductOrComment`

Reply:
Okay, I'm good with `ProductOrComment`.  I don't particularly like it, but the RFC doesn't give it a more sensible name.